src/lazy/encoder/binary/v1_1/value_writer.rs

```diff
@@ -231,8 +232,253 @@ impl<'value, 'top> BinaryValueWriter_1_1<'value, 'top> {
+                    let quarter_hours = ((offset_minutes - MIN_OFFSET) / 15) as u64;
+                    // A known offset uses 7 bits; opcode increases by 5, length (often) increases by 1.
+                    (7, quarter_hours, opcode_adjustment, length_adjustment)
+                }
+            };
+
+        // These opcodes assume UTC or unknown offset--we adjust for known offsets afterwards.
+        let (mut opcode, bits_populated, mut encoded_body_length, subseconds) =
+            match (precision, scale) {
+                (Year, _) => (0x70, 7, 1, 0),
+                (Month, _) => (0x71, 11, 2, 0),
+                (Day, _) => (0x72, 16, 2, 0),
+                (HourAndMinute, _) => (0x73, 27 + num_offset_bits, 4, 0),
+                // Seconds
+                (Second, 0) => (0x74, 33 + num_offset_bits, 5, 0),
+                // Milliseconds
+                (Second, 3) => (0x75, 43 + num_offset_bits, 6, value.milliseconds()),
+                // Microseconds
+                (Second, 6) => (0x76, 53 + num_offset_bits, 7, value.microseconds()),
+                // Nanoseconds
+                (Second, 9) => (0x77, 63 + num_offset_bits, 8, value.nanoseconds()),
+                _ => {
+                    unreachable!("illegal precision / fractional second seconds scale encountered")
+                }
+            };
+
+        // If the timestamp has a known offset, its opcode and encoded length must be adjusted.
+        // If it is unknown or UTC, the adjustment values will be zero making these steps into no-ops.
+        opcode += opcode_adjustment;
+        encoded_body_length += length_adjustment;
+
+        // Encode the offset
+        encoding |= offset_value << OFFSET_BIT_OFFSET;
+
+        // Because the seconds and subseconds follow the offset (which can be 1 bit or 7 bits) we need to dynamically
+        // calculate their respective bit offsets.
+        let seconds_bit_offset = OFFSET_BIT_OFFSET + num_offset_bits;
+        encoding |= (value.second() as u64) << seconds_bit_offset;
+
+        let subseconds_bit_offset = seconds_bit_offset + NUM_SECOND_BITS;
+        encoding |= (subseconds as u64) << subseconds_bit_offset;
+
+        // Because we eagerly (and branchless-ly) encoded all of the time units, we may have populated bits that are
+        // irrelevant to the final encoding. To simplify unit testing (and in the current absence of a binary 1.1
+        // reader), we calculate a mask of which bits are relevant to the current opcode and set any bits not in use
+        // to `0`.
+        // TODO: Remove this logic pending resolution of https://github.com/amazon-ion/ion-docs/issues/295, which
+        //       suggests requiring readers to ignore bits not used by the specified opcode
+        let mask = 1u64
+            .checked_shl(bits_populated)
+            .unwrap_or(0)
+            .checked_sub(1)
+            .unwrap_or(u64::MAX);
+        encoding &= mask;
```

Comment:
If we had a If we had a precomputed `const SHORT_TS_BIT_MASKS: [u64;65] = ...`, then this could be simplified to...
```rust
encoding &= SHORT_TS_BIT_MASKS[bits_populated];
```
...which might be cheap enough that we don't need to care about the performance aspect of ion-docs/295.

Reply:
I agree, but I'd like to leave this for a future PR.